### PR TITLE
Synchronize implementation of ECMA minimizers

### LINKED ISSUE #5454: Inconsistencies between ECMA and HTML mininfier
I was learning how to minify JS files from the [ECMA minifier CLI](https://github.com/swc-project/swc/blob/7437f53047853dad28d5b0274137a4167e74976c/crates/swc_ecma_minifier/src/cli/bin.rs). It uses all three transforms, the `resolver`, `hygiene` and `fixer`.

I discovered that the HTML minifier, [when it minimizes JS](https://github.com/swc-project/swc/blob/9d793f1cd3f2e23ba5b286783e0633a13690f1c2/crates/swc_html_minifier/src/lib.rs#L1452-L1571), only uses the `resolver` transform.

Is this intentional? Or a bug?

### PR DESCRIPTION
**Pull request by daxpedda**
**Description:**

This makes sure that all transforms are applied to all instances of ECMA minimization.
I also added `omit_trailing_semi`, I hope this is correct.

**Related issue (if exists):**
Fixes #5454.

### CODE REVIEW

File: crates/swc_html_minifier/src/lib.rs [hunk 1]
```diff
@@ -1533,17 +1533,25 @@ impl Minifier<'_> {
             },
         );
 
+        swc_ecma_visit::VisitMutWith::visit_mut_with(
```
Comment (kdy1) [inline crates/swc_html_minifier/src/lib.rs:1536]: We should not apply this

File: crates/swc_html_minifier/src/lib.rs [hunk 2]
```diff
@@ -1533,17 +1533,25 @@ impl Minifier<'_> {
...
-                cm.clone(),
-                "\n",
-                &mut buf,
-                None,
-            )) as Box<dyn swc_ecma_codegen::text_writer::WriteJs>;
-
-            wr = Box::new(swc_ecma_codegen::text_writer::omit_trailing_semi(wr));
+            let wr = swc_ecma_codegen::text_writer::JsWriter::new(cm.clone(), "\n", &mut buf, None);
+
+            let wr = Box::new(swc_ecma_codegen::text_writer::omit_trailing_semi(wr));
```
Comment (daxpedda) [inline crates/swc_html_minifier/src/lib.rs:1554]: Boxing it twice wasn't necessary.
Happy to clean this up by importing `JsWriter` and `omit_trailing_semi`? Was trying not to change too much.

File: crates/swc_ecma_minifier/examples/minifier.rs [hunk 3]
```diff
@@ -58,7 +58,7 @@ fn main() {
         )
         .expect_module();
 
-        let output = output.fold_with(&mut fixer(None));
+        let output = output.fold_with(&mut hygiene()).fold_with(&mut fixer(None));
```
Comment (kdy1) [inline crates/swc_ecma_minifier/examples/minifier.rs:61]: Please revery this

File: crates/swc_ecma_minifier/examples/minify-all.rs [hunk 4]
```diff
@@ -69,7 +69,7 @@ fn main() {
                         )
                         .expect_module();
 
-                        let output = output.fold_with(&mut fixer(None));
+                        let output = output.fold_with(&mut hygiene()).fold_with(&mut fixer(None));
```
Comment (kdy1) [inline crates/swc_ecma_minifier/examples/minify-all.rs:72]: Same here

File: crates/swc_ecma_minifier/examples/compress.rs [hunk 5]
```diff
@@ -59,7 +59,7 @@ fn main() {
         )
         .expect_module();
 
-        let output = output.fold_with(&mut fixer(None));
+        let output = output.fold_with(&mut hygiene()).fold_with(&mut fixer(None));
```
Comment (kdy1) [inline crates/swc_ecma_minifier/examples/compress.rs:62]: Same here

File: crates/swc_ecma_minifier/examples/minifier.rs (hunk 3 above)
Comment (daxpedda) [reply crates/swc_ecma_minifier/examples/minifier.rs:61]: So it should only apply fixer but not `hygiene`? I would like to understand why.

File: crates/swc_html_minifier/src/lib.rs (hunk 1 above)
Comment (daxpedda) [reply crates/swc_html_minifier/src/lib.rs:1536]: So it should only apply fixer but not `hygiene`? I would like to understand why.

Comment (daxpedda) [reply crates/swc_html_minifier/src/lib.rs:1536]: Done.

**Review by kdy1 — CHANGES_REQUESTED** [review]
We need `fixer`, and that was the issue

**Comment by daxpedda** [comment]
Should I remove `omit_trailing_semi`?
Should I make it in a separate PR or is it undesired? If yes, I would like to know why, should I not use it in my own code? Same with `hygiene`.

**Comment by kdy1** [comment]
No, hygiene and name mangler is playing same role, so it's not required for the minifier

**Comment by daxpedda** [comment]
Ah! Alright, thanks!

Is the PR alright like that, or should I also remove `omit_trailing_semi`?

**Comment by kdy1** [comment]
No, it's fine

**Pull request closed without merging**